File: .UNSAFE-RUST-DELETED-20260125/safe_float.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
// SPDX-FileCopyrightText: 2025 Hyperpolymath
//! Safe floating-point operations with NaN/Infinity prevention.
//!
//! This module provides mathematically verified safe operations for f32/f64,
//! preventing the common pitfalls that lead to NaN, Infinity, or crashes
//! in numerical computing and machine learning applications.
//!
//! # Design Philosophy
//!
//! Unlike integers which can overflow, floating-point numbers silently produce
//! NaN or Infinity on edge cases. This module makes those edge cases explicit
//! through Rust's Result type, following the proven library's pattern.
//!
//! # Example
//!
//! ```rust
//! use proven::SafeFloat;
//!
//! // Division by zero returns Error instead of Infinity
//! assert!(SafeFloat::div(1.0, 0.0).is_err());
//!
//! // Safe normalization prevents divide-by-zero on zero vectors
//! let v = vec![0.0, 0.0, 0.0];
//! assert!(SafeFloat::normalize(&v).is_err());
//!
//! // Safe log prevents NaN on non-positive values
//! assert!(SafeFloat::ln(0.0).is_err());
//! assert!(SafeFloat::ln(-1.0).is_err());
//! ```

use crate::core::{Error, Result};

/// Safe floating-point operations that never produce NaN or Infinity.
pub struct SafeFloat;

impl SafeFloat {
    /// Minimum positive value to consider non-zero (prevents denormal issues)
    pub const EPSILON: f64 = 1e-10;

    /// Safe division with zero check.
    ///
    /// Returns Error::DivisionByZero if divisor is zero or too small.
    ///
    /// # Example
    /// ```rust
    /// use proven::SafeFloat;
    ///
    /// assert_eq!(SafeFloat::div(10.0, 2.0).unwrap(), 5.0);
    /// assert!(SafeFloat::div(1.0, 0.0).is_err());
    /// ```
    pub fn div(a: f64, b: f64) -> Result<f64> {
        if b.abs() < Self::EPSILON {
            return Err(Error::DivisionByZero);
        }
        let result = a / b;
        if result.is_nan() || result.is_infinite() {
            return Err(Error::Overflow(format!("{} / {} is not finite", a, b)));
        }
        Ok(result)
    }

    /// Safe division for f32 with zero check.
    pub fn div_f32(a: f32, b: f32) -> Result<f32> {
        if b.abs() < f32::EPSILON {
            return Err(Error::DivisionByZero);
        }
        let result = a / b;
        if result.is_nan() || result.is_infinite() {
            return Err(Error::Overflow(format!("{} / {} is not finite", a, b)));
        }
        Ok(result)
    }

    /// Safe natural logarithm (ln).
    ///
    /// Returns Error for non-positive values (which would produce NaN or -Infinity).
    ///
    /// # Example
    /// ```rust
    /// use proven::SafeFloat;
    ///
    /// assert!(SafeFloat::ln(std::f64::consts::E).unwrap().abs() - 1.0 < 1e-10);
    /// assert!(SafeFloat::ln(0.0).is_err());
    /// assert!(SafeFloat::ln(-1.0).is_err());
    /// ```
    pub fn ln(x: f64) -> Result<f64> {
        if x <= 0.0 {
            return Err(Error::ValidationError(format!(
                "ln({}) undefined for non-positive values",
                x
            )));
        }
        Ok(x.ln())
    }

    /// Safe log base 10.
    pub fn log10(x: f64) -> Result<f64> {
        if x <= 0.0 {
            return Err(Error::ValidationError(format!(
                "log10({}) undefined for non-positive values",
                x
            )));
        }
        Ok(x.log10())
    }

    /// Safe square root.
    ///
    /// Returns Error for negative values.
    pub fn sqrt(x: f64) -> Result<f64> {
        if x < 0.0 {
            return Err(Error::ValidationError(format!(
                "sqrt({}) undefined for negative values",
                x
            )));
        }
        Ok(x.sqrt())
    }

    /// Safe power operation.
    ///
    /// Returns Error if result would be NaN or Infinity.
    pub fn pow(base: f64, exp: f64) -> Result<f64> {
        let result = base.powf(exp);
        if result.is_nan() || result.is_infinite() {
            return Err(Error::Overflow(format!(
                "{}^{} is not finite",
                base, exp
            )));
        }
        Ok(result)
    }

    /// Safe exponential (e^x).
    ///
    /// Returns Error if result would overflow to Infinity.
    pub fn exp(x: f64) -> Result<f64> {
        let result = x.exp();
        if result.is_infinite() {
            return Err(Error::Overflow(format!("exp({}) overflows", x)));
        }
        Ok(result)
    }

    /// Safe exponential for f32.
    pub fn exp_f32(x: f32) -> Result<f32> {
        let result = x.exp();
        if result.is_infinite() {
            return Err(Error::Overflow(format!("exp({}) overflows", x)));
        }
        Ok(result)
    }

    /// Compute vector magnitude safely.
    ///
    /// Returns the L2 norm (Euclidean length) of the vector.
    pub fn magnitude(v: &[f64]) -> f64 {
        v.iter().map(|x| x * x).sum::<f64>().sqrt()
    }

    /// Compute f32 vector magnitude safely.
    pub fn magnitude_f32(v: &[f32]) -> f32 {
        v.iter().map(|x| x * x).sum::<f32>().sqrt()
    }

    /// Safe vector normalization (unit vector).
    ///
    /// Returns Error if vector has zero magnitude.
    ///
    /// # Example
    /// ```rust
    /// use proven::SafeFloat;
    ///
    /// let v = vec![3.0, 4.0];
    /// let normalized = SafeFloat::normalize(&v).unwrap();
    /// assert!((normalized[0] - 0.6).abs() < 1e-10);
    /// assert!((normalized[1] - 0.8).abs() < 1e-10);
    ///
    /// // Zero vector cannot be normalized
    /// assert!(SafeFloat::normalize(&vec![0.0, 0.0]).is_err());
    /// ```
    pub fn normalize(v: &[f64]) -> Result<Vec<f64>> {
        let mag = Self::magnitude(v);
        if mag < Self::EPSILON {
            return Err(Error::DivisionByZero);
        }
        Ok(v.iter().map(|x| x / mag).collect())
    }

    /// Safe f32 vector normalization.
    pub fn normalize_f32(v: &[f32]) -> Result<Vec<f32>> {
        let mag = Self::magnitude_f32(v);
        if mag < f32::EPSILON {
            return Err(Error::DivisionByZero);
        }
        Ok(v.iter().map(|x| x / mag).collect())
    }

    /// Check if a float is finite (not NaN or Infinity).
    pub fn is_finite(x: f64) -> bool {
        x.is_finite()
    }

    /// Check if a float is safe for division (non-zero and finite).
    pub fn is_safe_divisor(x: f64) -> bool {
        x.is_finite() && x.abs() >= Self::EPSILON
    }

    /// Clamp a float to a range, handling NaN by returning min.
    ///
    /// Unlike std::f64::clamp which panics on NaN, this safely handles it.
    pub fn clamp(value: f64, min: f64, max: f64) -> f64 {
        if value.is_nan() {
            return min;
        }
        if value < min {
            min
        } else if value > max {
            max
        } else {
            value
        }
    }

    /// Safe reciprocal (1/x).
    pub fn reciprocal(x: f64) -> Result<f64> {
        Self::div(1.0, x)
    }

    /// Compute mean of a vector safely.
    ///
    /// Returns Error for empty vectors.
    pub fn mean(v: &[f64]) -> Result<f64> {
        if v.is_empty() {
            return Err(Error::EmptyInput);
        }
        let sum: f64 = v.iter().sum();
        Self::div(sum, v.len() as f64)
    }

    /// Compute mean of f32 vector.
    pub fn mean_f32(v: &[f32]) -> Result<f32> {
        if v.is_empty() {
            return Err(Error::EmptyInput);
        }
        let sum: f32 = v.iter().sum();
        Self::div_f32(sum, v.len() as f32)
    }

    /// Compute variance safely.
    pub fn variance(v: &[f64]) -> Result<f64> {
        let mean = Self::mean(v)?;
        let sum_sq: f64 = v.iter().map(|x| (x - mean).powi(2)).sum();
        Self::div(sum_sq, v.len() as f64)
    }

    /// Compute standard deviation safely.
    pub fn std_dev(v: &[f64]) -> Result<f64> {
        let var = Self::variance(v)?;
        Self::sqrt(var)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_div_normal() {
        assert_eq!(SafeFloat::div(10.0, 2.0).unwrap(), 5.0);
        assert_eq!(SafeFloat::div(-10.0, 2.0).unwrap(), -5.0);
    }

    #[test]
    fn test_div_by_zero() {
        assert!(SafeFloat::div(1.0, 0.0).is_err());
        assert!(SafeFloat::div(0.0, 0.0).is_err());
        assert!(SafeFloat::div_f32(1.0, 0.0).is_err());
    }

    #[test]
    fn test_ln_positive() {
        let result = SafeFloat::ln(std::f64::consts::E).unwrap();
        assert!((result - 1.0).abs() < 1e-10);
    }

    #[test]
    fn test_ln_non_positive() {
        assert!(SafeFloat::ln(0.0).is_err());
        assert!(SafeFloat::ln(-1.0).is_err());
    }

    #[test]
    fn test_sqrt_positive() {
        assert_eq!(SafeFloat::sqrt(4.0).unwrap(), 2.0);
        assert_eq!(SafeFloat::sqrt(0.0).unwrap(), 0.0);
    }

    #[test]
    fn test_sqrt_negative() {
        assert!(SafeFloat::sqrt(-1.0).is_err());
    }

    #[test]
    fn test_normalize() {
        let v = vec![3.0, 4.0];
        let n = SafeFloat::normalize(&v).unwrap();
        assert!((n[0] - 0.6).abs() < 1e-10);
        assert!((n[1] - 0.8).abs() < 1e-10);

        // Magnitude should be 1.0
        let mag = SafeFloat::magnitude(&n);
        assert!((mag - 1.0).abs() < 1e-10);
    }

    #[test]
    fn test_normalize_zero_vector() {
        let v = vec![0.0, 0.0, 0.0];
        assert!(SafeFloat::normalize(&v).is_err());
    }

    #[test]
    fn test_exp_normal() {
        assert!((SafeFloat::exp(0.0).unwrap() - 1.0).abs() < 1e-10);
        assert!((SafeFloat::exp(1.0).unwrap() - std::f64::consts::E).abs() < 1e-10);
    }

    #[test]
    fn test_exp_overflow() {
        assert!(SafeFloat::exp(1000.0).is_err());
    }

    #[test]
    fn test_clamp_nan() {
        let nan = f64::NAN;
        assert_eq!(SafeFloat::clamp(nan, 0.0, 1.0), 0.0);
    }

    #[test]
    fn test_clamp_normal() {
        assert_eq!(SafeFloat::clamp(0.5, 0.0, 1.0), 0.5);
        assert_eq!(SafeFloat::clamp(-1.0, 0.0, 1.0), 0.0);
        assert_eq!(SafeFloat::clamp(2.0, 0.0, 1.0), 1.0);
    }

    #[test]
    fn test_mean() {
        let v = vec![1.0, 2.0, 3.0, 4.0, 5.0];
        assert_eq!(SafeFloat::mean(&v).unwrap(), 3.0);
    }

    #[test]
    fn test_mean_empty() {
        let v: Vec<f64> = vec![];
        assert!(SafeFloat::mean(&v).is_err());
    }

    #[test]
    fn test_variance() {
        let v = vec![2.0, 4.0, 4.0, 4.0, 5.0, 5.0, 7.0, 9.0];
        let var = SafeFloat::variance(&v).unwrap();
        assert!((var - 4.0).abs() < 1e-10);
    }

    #[test]
    fn test_std_dev() {
        let v = vec![2.0, 4.0, 4.0, 4.0, 5.0, 5.0, 7.0, 9.0];
        let sd = SafeFloat::std_dev(&v).unwrap();
        assert!((sd - 2.0).abs() < 1e-10);
    }
}

File: bindings/r/R/safe_float.R
# SPDX-License-Identifier: Apache-2.0
# SPDX-FileCopyrightText: 2025 Hyperpolymath

#' Safe floating-point operations with NaN/Infinity prevention.
#'
#' Provides mathematically verified safe operations that prevent
#' NaN, Infinity, or crashes in numerical computing.

#' Minimum positive value to consider non-zero.
#' @export
FLOAT_EPSILON <- 1e-10

#' Safe division with zero check.
#'
#' Returns NA if divisor is zero or too small.
#'
#' @param a Dividend
#' @param b Divisor
#' @return Quotient or NA on error
#' @export
safe_div <- function(a, b) {
  if (is.na(a) || is.na(b)) return(NA_real_)
  if (abs(b) < FLOAT_EPSILON) return(NA_real_)

  result <- a / b
  if (!is.finite(result)) return(NA_real_)

  result
}

#' Safe natural logarithm.
#'
#' Returns NA for non-positive values.
#'
#' @param x Value
#' @return Natural log or NA on error
#' @export
safe_ln <- function(x) {
  if (is.na(x) || x <= 0) return(NA_real_)
  log(x)
}

#' Safe log base 10.
#'
#' Returns NA for non-positive values.
#'
#' @param x Value
#' @return Log base 10 or NA on error
#' @export
safe_log10 <- function(x) {
  if (is.na(x) || x <= 0) return(NA_real_)
  log10(x)
}

#' Safe square root.
#'
#' Returns NA for negative values.
#'
#' @param x Value
#' @return Square root or NA on error
#' @export
safe_sqrt <- function(x) {
  if (is.na(x) || x < 0) return(NA_real_)
  sqrt(x)
}

#' Safe power operation.
#'
#' Returns NA if result is not finite.
#'
#' @param base Base value
#' @param exp Exponent
#' @return Power or NA on error
#' @export
safe_pow <- function(base, exp) {
  if (is.na(base) || is.na(exp)) return(NA_real_)

  result <- base ^ exp
  if (!is.finite(result)) return(NA_real_)

  result
}

#' Safe exponential (e^x).
#'
#' Returns NA if result overflows.
#'
#' @param x Exponent
#' @return Exponential or NA on overflow
#' @export
safe_exp <- function(x) {
  if (is.na(x)) return(NA_real_)

  result <- exp(x)
  if (!is.finite(result)) return(NA_real_)

  result
}

#' Compute vector magnitude (L2 norm).
#'
#' @param v Numeric vector
#' @return Magnitude
#' @export
float_magnitude <- function(v) {
  if (length(v) == 0) return(0)
  sqrt(sum(v^2))
}

#' Safe vector normalization (unit vector).
#'
#' Returns NA if vector has zero magnitude.
#'
#' @param v Numeric vector
#' @return Normalized vector or NA on error
#' @export
float_normalize <- function(v) {
  if (length(v) == 0) return(NA)

  mag <- float_magnitude(v)
  if (mag < FLOAT_EPSILON) return(NA)

  v / mag
}

#' Check if a float is finite (not NaN or Infinity).
#'
#' @param x Value to check
#' @return TRUE if finite
#' @export
float_is_finite <- function(x) {
  !is.na(x) && is.finite(x)
}

#' Check if a float is safe for division.
#'
#' @param x Value to check
#' @return TRUE if safe divisor
#' @export
float_is_safe_divisor <- function(x) {
  float_is_finite(x) && abs(x) >= FLOAT_EPSILON
}

#' Clamp a float to a range, handling NaN.
#'
#' Unlike base clamp, this safely handles NaN by returning min.
#'
#' @param value Value to clamp
#' @param min_val Minimum value
#' @param max_val Maximum value
#' @return Clamped value
#' @export
float_clamp <- function(value, min_val, max_val) {
  if (is.na(value) || !is.finite(value)) return(min_val)
  if (value < min_val) return(min_val)
  if (value > max_val) return(max_val)
  value
}

#' Safe reciprocal (1/x).
#'
#' @param x Value
#' @return Reciprocal or NA on error
#' @export
safe_reciprocal <- function(x) {
  safe_div(1, x)
}

#' Compute mean of a vector safely.
#'
#' Returns NA for empty vectors.
#'
#' @param v Numeric vector
#' @return Mean or NA on error
#' @export
safe_mean <- function(v) {
  if (length(v) == 0) return(NA_real_)
  if (any(is.na(v))) return(NA_real_)

  sum(v) / length(v)
}

#' Compute variance safely.
#'
#' @param v Numeric vector
#' @return Variance or NA on error
#' @export
safe_variance <- function(v) {
  mean_val <- safe_mean(v)
  if (is.na(mean_val)) return(NA_real_)

  sum_sq <- sum((v - mean_val)^2)
  safe_div(sum_sq, length(v))
}

#' Compute standard deviation safely.
#'
#' @param v Numeric vector
#' @return Standard deviation or NA on error
#' @export
safe_std_dev <- function(v) {
  var_val <- safe_variance(v)
  if (is.na(var_val)) return(NA_real_)

  safe_sqrt(var_val)
}

#' Safe sigmoid function.
#'
#' @param x Input value
#' @return Sigmoid output (0-1) or NA on error
#' @export
safe_sigmoid <- function(x) {
  if (is.na(x)) return(NA_real_)

  # Prevent overflow in exp
  if (x < -700) return(0)
  if (x > 700) return(1)

  1 / (1 + exp(-x))
}

#' Safe softmax function.
#'
#' @param v Numeric vector
#' @return Softmax probabilities or NA on error
#' @export
safe_softmax <- function(v) {
  if (length(v) == 0) return(NA)
  if (any(is.na(v))) return(NA)

  # Subtract max for numerical stability
  v_shifted <- v - max(v)
  exp_v <- exp(v_shifted)
  total <- sum(exp_v)

  if (total < FLOAT_EPSILON) return(NA)

  exp_v / total
}

#' Safe log-sum-exp.
#'
#' @param v Numeric vector
#' @return Log-sum-exp or NA on error
#' @export
safe_logsumexp <- function(v) {
  if (length(v) == 0) return(NA_real_)
  if (any(is.na(v))) return(NA_real_)

  max_v <- max(v)
  max_v + log(sum(exp(v - max_v)))
}

#' Compute dot product safely.
#'
#' @param a First vector
#' @param b Second vector
#' @return Dot product or NA if lengths differ
#' @export
safe_dot_product <- function(a, b) {
  if (length(a) != length(b)) return(NA_real_)
  if (length(a) == 0) return(0)
  if (any(is.na(a)) || any(is.na(b))) return(NA_real_)

  sum(a * b)
}

#' Compute cosine similarity safely.
#'
#' @param a First vector
#' @param b Second vector
#' @return Cosine similarity (-1 to 1) or NA on error
#' @export
safe_cosine_similarity <- function(a, b) {
  if (length(a) != length(b)) return(NA_real_)

  dot <- safe_dot_product(a, b)
  if (is.na(dot)) return(NA_real_)

  mag_a <- float_magnitude(a)
  mag_b <- float_magnitude(b)

  if (mag_a < FLOAT_EPSILON || mag_b < FLOAT_EPSILON) {
    return(NA_real_)
  }

  float_clamp(dot / (mag_a * mag_b), -1, 1)
}

#' Compute Euclidean distance safely.
#'
#' @param a First vector
#' @param b Second vector
#' @return Distance or NA on error
#' @export
safe_euclidean_distance <- function(a, b) {
  if (length(a) != length(b)) return(NA_real_)
  if (any(is.na(a)) || any(is.na(b))) return(NA_real_)

  sqrt(sum((a - b)^2))
}

#' Linear interpolation (lerp).
#'
#' @param a Start value
#' @param b End value
#' @param t Interpolation factor (0-1)
#' @return Interpolated value
#' @export
float_lerp <- function(a, b, t) {
  if (is.na(a) || is.na(b) || is.na(t)) return(NA_real_)
  t <- float_clamp(t, 0, 1)
  a + (b - a) * t
}

#' Check if two floats are approximately equal.
#'
#' @param a First value
#' @param b Second value
#' @param epsilon Tolerance (default FLOAT_EPSILON)
#' @return TRUE if approximately equal
#' @export
float_approx_equal <- function(a, b, epsilon = FLOAT_EPSILON) {
  if (is.na(a) || is.na(b)) return(FALSE)
  abs(a - b) < epsilon
}
